### PR TITLE
Proof Verifier Integration (Cumulative/Multi Sidechain) aka Darlin

### PR DESCRIPTION
**Pull request by abi87**
Refactored code as follows:

1. Moved mempool processing to different thread, while encapsulating some global state into ad-hoc manager
2. Broken mempool processing into sc proof verifications vs all the rest, to be able to selectively do one check or the other
3. Batched proof verification across certs and csw/mbtr txes.

Multithreading structure may be revaluated once the actual proof implementation is integrated

### CODE REVIEW

File: src/txBaseMsgProcessor.cpp [hunk 1]
```diff
@@ -0,0 +1,403 @@
...
+    return;
+}
+
+void TxBaseMsgProcessor::addTxBaseMsgToProcess(const CTransactionBase& txBase, CNodeInterface* pfrom)
+{
+    CInv inv(MSG_TX, txBase.GetHash());
+    pfrom->AddInventoryKnown(inv);
+
+    pfrom->StopAskingFor(inv);
+    mapAlreadyAskedFor.erase(inv);
```
Comment (i-Alex) [inline src/txBaseMsgProcessor.cpp:44]: In previous implementation `mapAlreadyAskedFor.erase(inv)` was executed with lock `LOCK(cs_main);`
Please, check the usage for possible conflicts.

File: src/main.cpp [hunk 2]
```diff
@@ -4898,11 +4936,8 @@ bool LoadBlockIndex()
 
 
 bool InitBlockIndex() {
-    const CChainParams& chainparams = Params();
     LOCK(cs_main);
-
-    // Initialize global variables that cannot be constructed at startup.
-    recentRejects.reset(new CRollingBloomFilter(120000, 0.000001));
+    const CChainParams& chainparams = Params();
```
Comment (alsala) [inline src/main.cpp:4940]: The call to Params() now is in the critical section, probably it is not needed

File: src/txBaseMsgProcessor.cpp [hunk 3]
```diff
@@ -0,0 +1,482 @@
...
+    assert(recentRejects);
+    if (currentTipHash != hashRecentRejectsChainTip)
+    {
+        // If the chain tip has changed previously rejected transactions
+        // might be now valid, e.g. due to a nLockTime'd tx becoming valid,
+        // or a double-spend. Reset the rejects filter and give those
+        // txs a second chance.
+        hashRecentRejectsChainTip = currentTipHash;
+        recentRejects.reset(nullptr);
+        recentRejects.reset(new CRollingBloomFilter(120000, 0.000001));
```
Comment (alsala) [inline src/txBaseMsgProcessor.cpp:479]: This seems correct but it is not clear to me: in the original code there were just a call to reset() in this point the obj was not reset to a new filter.

File: src/main.cpp [hunk 4]
```diff
@@ -5305,27 +5340,18 @@ std::string GetWarnings(const std::string& strFor)
 //
 
 
-bool static AlreadyHave(const CInv& inv) EXCLUSIVE_LOCKS_REQUIRED(cs_main)
+bool AlreadyHave(const CInv& inv)
```
Comment (alsala) [inline src/main.cpp:5343]: This function is still called with the cs_main lock acquired, consider restoring the attribute

Comment (abi87) [reply src/main.cpp:5343]: Done in main.h

File: src/main.cpp [hunk 5]
```diff
@@ -2902,13 +2923,15 @@ bool ConnectBlock(const CBlock& block, CValidationState& state, CBlockIndex* pin
...
 
-        auto scVerifier = fExpensiveChecks ? libzendoomc::CScProofVerifier::Strict() : libzendoomc::CScProofVerifier::Disabled();
-        if (!view.IsCertApplicableToState(cert, scVerifier) )
+        if (!view.IsCertApplicableToStateWithoutProof(cert))
         {
             return state.DoS(100, error("%s():%d: invalid sc certificate [%s]", cert.GetHash().ToString(),__func__, __LINE__),
                              REJECT_INVALID, "bad-sc-cert-not-applicable");
         }
 
+        if (fScRelatedChecks)
```
[the remark this answers is not in the document]
Comment (abi87) [reply src/main.cpp:2932]: Agree. Let's leave this comment open for future reference

File: src/main.cpp [hunk 6]
```diff
@@ -2802,11 +2796,16 @@ bool ConnectBlock(const CBlock& block, CValidationState& state, CBlockIndex* pin
...
             {
                 return state.DoS(100, error("%s():%d - ERROR: tx=%s\n", __func__, __LINE__, tx.GetHash().ToString()),
-                                 REJECT_INVALID, "bad-sc-tx");
+                                 REJECT_INVALID, "bad-sc-tx-not-applicable");
+            }
+
+            if (fScRelatedChecks)
+            {
+                scVerifier.LoadDataForMbtrVerification(view, tx);
+                scVerifier.LoadDataForCswVerification(view, tx);
```
[the remark this answers is not in the document]
Comment (abi87) [reply src/main.cpp:2808]: I actually chose on purpose to give maximal granularity to these methods, anticipating the drop of proof in mbtr. Hope i paid off. Anyhow I would keep these as is and deal with interface once merge with https://github.com/HorizenOfficial/zend_oo/pull/105

File: src/txBaseMsgProcessor.h [hunk 7]
```diff
@@ -0,0 +1,116 @@
...
+        // data following tx processing, needed in case a re-processing will occur (e.g. partial validation)
+        MempoolReturnValue txBaseProcessingState;
+        CValidationState   txBaseValidationState;
+
+        TxBaseMsg_DataToProcess();
+    };
+
+    mutable boost::mutex mutex;
+    mutable boost::condition_variable condWorker;
+    std::vector<TxBaseMsg_DataToProcess> txBaseMsgQueue GUARDED_BY(cs_txesUnderProcess);
```
[the remark this answers is not in the document]
Comment (abi87) [reply src/txBaseMsgProcessor.h:113]: Indeed, updated

File: src/txBaseMsgProcessor.cpp (hunk 3 above)
Comment (abi87) [reply src/txBaseMsgProcessor.cpp:479]: The previous reset used to wipe the `CRollingBloomFilter` out, while another, not encapsulated call would instantiate the filter. Here I have encapsulated both these operations resulting in the double reset.
I agree it's not pretty, but it allows me to have a single reset to be called anywhere

File: src/main.cpp [hunk 8]
```diff
@@ -4859,14 +4898,14 @@ bool CVerifyDB::VerifyDB(CCoinsView *coinsview, int nCheckLevel, int nCheckDepth
 
 void UnloadBlockIndex()
 {
+    TxBaseMsgProcessor::get().reset();
```
[the remark this answers is not in the document]
Comment (abi87) [reply src/main.cpp:4901]: Done

File: src/main.cpp [hunk 9]
```diff
@@ -1133,39 +1020,57 @@ CAmount GetMinRelayFee(const CTransactionBase& tx, unsigned int nBytes, bool fAl
     return nMinFee;
 }
 
-bool AcceptCertificateToMemoryPool(CTxMemPool& pool, CValidationState &state, const CScCertificate &cert,
-    LimitFreeFlag fLimitFree, bool* pfMissingInputs, RejectAbsurdFeeFlag fRejectAbsurdFee)
+void StoreCertToMempool(const CScCertificate &cert, CTxMemPool &pool, const CCoinsViewCache &view)
+{
+    std::pair<uint256, CAmount> conflictingCertData = pool.FindCertWithQuality(cert.GetScId(), cert.quality);
+    pool.RemoveCertAndSync(conflictingCertData.first);
```
[the remark this answers is not in the document]
Comment (abi87) [reply src/main.cpp:1028]: great catch! I have introduce some rechecks, specifically related to mempool/view to make sure that a transaction is still valid before attempting the batched verification

File: src/txBaseMsgProcessor.cpp (hunk 1 above)
Comment (abi87) [reply src/txBaseMsgProcessor.cpp:39]: It's fine, I believe previous usage was an overkill. I also re-run some integration tests with the thread-sanitizer on and got no issues (at least with the specific executions I performed).

File: src/sc/sidechain.cpp [hunk 10]
```diff
@@ -12,7 +11,44 @@
...
+static const boost::filesystem::path Sidechain::GetSidechainDataDir()
+{
+    static const boost::filesystem::path sidechainsDataDir = GetDataDir() / "sidechains";
+    return sidechainsDataDir;
+}
+
+bool Sidechain::InitSidechainsFolder()
+{
+    // Note: sidechainsDataDir cannot be global since
+    // at start of the program network parameters are not initilized yet
```
[the remark this answers is not in the document]
Comment (abi87) [reply src/sc/sidechain.cpp]: Done

File: src/miner.cpp [hunk 11]
```diff
@@ -760,7 +760,7 @@ CBlockTemplate* CreateNewBlock(const CScript& scriptPubKeyIn,  unsigned int nBlo
         pblocktemplate->vTxSigOps[0] = GetLegacySigOpCount(pblock->vtx[0]);
 
         CValidationState state;
-        if (!TestBlockValidity(state, *pblock, pindexPrev, false, false, false))
+        if (!TestBlockValidity(state, *pblock, pindexPrev, /*fCheckPOW*/false, /*fCheckMerkleRoot*/false, /*fScRelatedChecks*/false))
```
[the remark this answers is not in the document]
Comment (abi87) [reply src/miner.cpp]: Done

File: src/gtest/test_processTxBaseMsg.cpp [hunk 12]
```diff
@@ -0,0 +1,831 @@
...
+{
+public:
+    FakeNode(): fWhitelisted(false) {};
+    ~FakeNode() = default;
+
+    bool fWhitelisted;
+    std::string commandInvoked;
+
+    void AddInventoryKnown(const CInv& inv)  {}; //dummyImpl
+    NodeId GetId() const  {return 1987;};
```
[the remark this answers is not in the document]
Comment (abi87) [reply src/gtest/test_processTxBaseMsg.cpp]: Done

File: src/primitives/transaction.h [hunk 13]
```diff
@@ -674,18 +671,7 @@ class CBwtRequestOut : public CTxCrosschainOutBase
...
 class CValidationState;
-class CTxMemPool;
-class CCoinsViewCache;
 class CChain;
-class CBlock;
-class CBlockTemplate;
-class CScriptCheck;
-class CBlockUndo;
-class CTxUndo;
-class UniValue;
```
[the remark this answers is not in the document]
Comment (abi87) [reply src/primitives/transaction.h:685]: Indeed `#include <univalue.h>` can be removed while fwd declarations `CTxUndo`,  `CBlockUndo` and `CScriptCheck` are not necessary.
Moved up in file the other needed forward declarations

File: src/txBaseMsgProcessor.cpp [hunk 14]
```diff
@@ -208,43 +203,135 @@ void TxBaseMsgProcessor::ProcessTxBaseMsg(const processMempoolTx& mempoolProcess
...
+            // update in-place all PARTIALLY_VALIDATED txes/certs
+            txBaseMsgQueue.push_back(dataToProcess); //Reinsert dataToProcess to process all PARTIALLY_VALIDATED txes
+            for(unsigned int idx = 0; idx < txBaseMsgQueue.size(); ++idx)
+            {
+                TxBaseMsg_DataToProcess& enqueuedItem = txBaseMsgQueue.at(idx);
+                if (enqueuedItem.txBaseProcessingState != MempoolReturnValue::PARTIALLY_VALIDATED)
+                    continue;
+
+                // Mempool may have changed since partial validation has been made. Recheck for double spends
+                if (!supportView.HaveInputs(*enqueuedItem.pTxBase))
```
Comment (i-Alex) [inline src/txBaseMsgProcessor.cpp:221]: We check Certs and Txs from the queue against `supportView`, but never update the `supportView`. is it fine?
Are we sure that in the queue there are no txs/certs spending the same inputs?

Maybe we need to check this after proof validation before running `StoreCertToMempool/StoreTxToMempool`.

File: src/main.cpp [hunk 15]
```diff
@@ -1502,11 +1508,30 @@ MempoolReturnValue AcceptTxToMemoryPool(CTxMemPool& pool, CValidationState &stat
...
+        if (resMbtr.count(hash) != 0 && !resMbtr.at(hash))
+        {
+            state.Invalid(error("%s():%d - ERROR: sc-related tx [%s] proofs do not verify\n",
+                          __func__, __LINE__, hash.ToString()), REJECT_INVALID, "bad-sc-tx-proof");
+            return MempoolReturnValue::INVALID;
+        }
+
+        scVerifier.LoadDataForCswVerification(view, tx);
+        std::map<uint256, bool> resCsw = scVerifier.batchVerifyCsws();
+        if (resCsw.count(hash) != 0 && !resCsw.at(hash))
```
Comment (i-Alex) [inline src/main.cpp:1534]: maybe `if (resCsw.count(hash) == 0 || !resCsw.at(hash))`
to cover the potential issue caused by scVerifer returning empty `resCsw` or `resCsw` with no `hash`.

Same for mbtrs.

File: src/main.cpp (hunk 6 above)
[the remark this answers is not in the document]
Comment (i-Alex) [reply src/main.cpp:2810]: Agree with Albene. Anyway we will remove `LoadDataForMbtrVerification` logic, because there will be no proofs for mbtrs.

**Review by i-Alex — CHANGES_REQUESTED** [review]

File: src/main.cpp (hunk 15 above)
Comment (abi87) [reply src/main.cpp:1534]: Actually `resCsw.count(hash) == 0` covers the case of a tx with bwts but not csw. If I replace `&&` with `||` I would flag all such txes as invalid. We may implement your suggestion for bwts once we remove proof for csws

File: src/sc/proofverifier.cpp [hunk 16]
```diff
@@ -1,97 +1,205 @@
...
+        assert(view.GetSidechain(csw.scId, sidechain) && "Unknown sidechain at scTx proof verification stage");
+
+        cswVerifierInputsList& newItem = cswEnqueuedData[scTx.GetHash()][idx]; //create or retrieve new entry
+        newItem.certDataHash = view.GetActiveCertDataHash(csw.scId);
+//        //TODO: Unlock when we'll handle recovery of fwt of last epoch
+//        if (certDataHash.IsNull())
+//            return error("%s():%d - ERROR: Tx[%s] CSW input [%s] has missing active cert data hash for required scId[%s]\n",
+//                            __func__, __LINE__, tx.ToString(), csw.ToString(), csw.scId.ToString());
+
+        if (sidechain.creationData.wCeasedVk.is_initialized())
```
[the remark this answers is not in the document]
Comment (abi87) [reply src/sc/proofverifier.cpp:184]: Agree, done

File: src/txBaseMsgProcessor.cpp [hunk 17]
```diff
@@ -0,0 +1,444 @@
...
+
+TxBaseMsgProcessor::TxBaseMsgProcessor() { this->reset(); }
+TxBaseMsgProcessor::~TxBaseMsgProcessor() { this->reset(); }
+
+void TxBaseMsgProcessor::reset()
+{
+    {
+        LOCK(cs_rejectedTxes);
+        recentRejects.reset(nullptr);
+        recentRejects.reset(new CRollingBloomFilter(120000, 0.000001));
```
Comment (ptagl) [inline src/txBaseMsgProcessor.cpp:22]: It seems like the first reset (with _nullptr_) is not necessary.

```suggestion
        recentRejects.reset(new CRollingBloomFilter(120000, 0.000001));
```

File: src/txBaseMsgProcessor.cpp [hunk 18]
```diff
@@ -0,0 +1,444 @@
...
+    assert(recentRejects);
+    if (currentTipHash != hashRecentRejectsChainTip)
+    {
+        // If the chain tip has changed previously rejected transactions
+        // might be now valid, e.g. due to a nLockTime'd tx becoming valid,
+        // or a double-spend. Reset the rejects filter and give those
+        // txs a second chance.
+        hashRecentRejectsChainTip = currentTipHash;
+        recentRejects.reset(nullptr);
+        recentRejects.reset(new CRollingBloomFilter(120000, 0.000001));
```
Comment (ptagl) [inline src/txBaseMsgProcessor.cpp:441]: It seems like the first reset (with _nullptr_) is not necessary.

```suggestion
        recentRejects.reset(new CRollingBloomFilter(120000, 0.000001));
```

File: src/txBaseMsgProcessor.h [hunk 19]
```diff
@@ -0,0 +1,112 @@
...
+     * two minute window to send invs to us.
+     *
+     * Decreasing the false positive rate is fairly cheap, so we pick one in a
+     * million to make it highly unlikely for users to have issues with this
+     * filter.
+     *
+     * Memory used: 1.7MB
+     */
+    mutable CCriticalSection cs_rejectedTxes;
+    boost::scoped_ptr<CRollingBloomFilter> recentRejects  GUARDED_BY(cs_rejectedTxes);
```
Comment (ptagl) [inline src/txBaseMsgProcessor.h:88]: If there is no particular reason behind it, I would suggest to switch from `boost::scoped_ptr` to `std::unique_ptr` to reduce unnecessary dependency from Boost library.

```suggestion
    std::unique_ptr<CRollingBloomFilter> recentRejects  GUARDED_BY(cs_rejectedTxes);
```

File: qa/rpc-tests/sc_cert_base.py [hunk 20]
```diff
@@ -250,7 +250,7 @@ def run_test(self):
             errorString = e.error['message']
             mark_logs(errorString, self.nodes, DEBUG_MODE)
 
-        assert_equal("bad-sc-cert-not-applicable" in errorString, True)
+        assert_equal("bad-sc-cert-proof" in errorString, True)
```
Comment (ptagl) [inline qa/rpc-tests/sc_cert_base.py:253]: Since `bad-sc-cert-proof` is used several time, consider storing it in a (const) variable.

**Comment by ptagl** [comment]
This PR is now obsolete.

**Pull request closed without merging**